tidy3d/plugins/invdes/optimizer.py

```diff
@@ -0,0 +1,175 @@
+        print(f"step ({step_index + 1}/{self.num_steps})")
+        print(f"\tobjective_fn_val = {result.objective_fn_val[-1]:.3e}")
+        print(f"\tgrad_norm = {jnp.linalg.norm(result.grad[-1]):.3e}")
+        print(f"\tpost_process_val = {result.post_process_val[-1]:.3e}")
+        print(f"\tpenalty = {result.penalty[-1]:.3e}")
+
+    def _initialize_result(self, params0: jnp.ndarray) -> InverseDesignResult:
+        """Create an initially empty ``InverseDesignResult`` from the starting parameters."""
+
+        # initialize optimizer
+        params0 = jnp.array(params0)
+        optax_optimizer = self.optax_optimizer
+        opt_state = optax_optimizer.init(params0)
+
+        # initialize empty result
+        return InverseDesignResult(design=self.design, opt_state=[opt_state], params=[params0])
+
+    def run(self, params0: jnp.ndarray) -> InverseDesignResult:
+        """Run this inverse design problem from an initial set of parameters."""
+        self.design.design_region._check_params(params0)
+        starting_result = self._initialize_result(params0)
+        return self.continue_run(result=starting_result)
+
+    def continue_run(self, result: InverseDesignResult) -> InverseDesignResult:
+        """Run optimizer for a series of steps with an initialized state."""
+
+        # get the last state of the optimizer and the last number of params
+        opt_state = result.get_final("opt_state")
+        params = result.get_final("params")
+        history = deepcopy(result.history)
+
+        # use jax to grad the objective function
+        objective_fn = self.design.objective_fn
+        val_and_grad_fn = jax.value_and_grad(objective_fn, has_aux=True)
+
+        optax_optimizer = self.optax_optimizer
+
+        # main optimization loop
+        for step_index in range(self.num_steps):
+            # evaluate gradient
+
+            (val, aux_data), grad = val_and_grad_fn(params, step_index=step_index, history=history)
+            # TODO: add more kwargs here?
+
+            # strip out auxiliary data
+            penalty = aux_data["penalty"]
+            post_process_val = aux_data["post_process_val"]
+
+            # save history
+            history["objective_fn_val"].append(val)
+            history["grad"].append(grad)
+            history["penalty"].append(penalty)
+            history["post_process_val"].append(post_process_val)
+
+            # display information
+            result = InverseDesignResult(design=result.design, **history)
+            self.display_fn(result, step_index=step_index)
+
+            # update optimizer and parameters
+            updates, opt_state = optax_optimizer.update(-grad, opt_state, params)
+            params = optax.apply_updates(params, updates)
+
+            # cap the parameters
+            params = jnp.minimum(params, 1.0)
+            params = jnp.maximum(params, 0.0)
+
+            history["params"].append(params)
+            history["opt_state"].append(opt_state)
```

Comment:
So what I did is when the optimizer is run from an empty state `Optimizer.run(params)`, it will store the initial `params` and `opt_state` before running this loop. 

https://github.com/flexcompute/tidy3d/blob/18449686c9fe520138170792b2b108684d4eb9d8/tidy3d/plugins/invdes/optimizer.py#L76

So actually for `num_steps` steps, we end up having `num_steps+1` values for `params` and `opt_state` in the history but only `num_steps` values for the other stuff.

It's not always clear whether people want the final state / parameters to be **after** the final update or what. But I decided to include it here just for completeness. And also so that if we continue a run it will not waste that gradient calculation.